nexus/reconfigurator/planning/src/planner.rs

```diff
@@ -2180,4 +2204,421 @@ mod test {
+            .filter(|z| z.zone_type.is_clickhouse_server())
+            .map(|z| z.id)
+            .collect();
+
+        assert_eq!(keeper_zone_ids.len(), target_keepers);
+        assert_eq!(server_zone_ids.len(), target_servers);
+
+        // Directly manipulate the blueprint and inventory so that the
+        // clickhouse clusters are stable
+        let config = blueprint2.clickhouse_cluster_config.as_mut().unwrap();
+        config.max_used_keeper_id = (target_keepers as u64).into();
+        config.keepers = keeper_zone_ids
+            .iter()
+            .enumerate()
+            .map(|(i, zone_id)| (*zone_id, KeeperId(i as u64)))
+            .collect();
+        config.highest_seen_keeper_leader_committed_log_index = 1;
+
+        let raft_config: BTreeSet<_> =
+            config.keepers.values().cloned().collect();
+
+        collection.clickhouse_keeper_cluster_membership = config
+            .keepers
+            .iter()
+            .map(|(zone_id, keeper_id)| {
+                (
+                    *zone_id,
+                    ClickhouseKeeperClusterMembership {
+                        queried_keeper: *keeper_id,
+                        leader_committed_log_index: 1,
+                        raft_config: raft_config.clone(),
+                    },
+                )
+            })
+            .collect();
+
+        // Let's run the planner. The blueprint shouldn't change with regards to
+        // clickhouse as our inventory reflects our desired state.
+        let blueprint3 = Planner::new_based_on(
+            log.clone(),
+            &blueprint2,
+            &input,
+            "test_blueprint3",
+            &collection,
+        )
+        .expect("created planner")
+        .with_rng_seed((TEST_NAME, "bp3"))
+        .plan()
+        .expect("plan");
+
+        assert_eq!(
+            blueprint2.clickhouse_cluster_config,
+            blueprint3.clickhouse_cluster_config
+        );
+
+        // Find the sled containing one of the keeper zones and expunge it
+        let (sled_id, bp_zone_config) = blueprint3
+            .all_omicron_zones(BlueprintZoneFilter::ShouldBeRunning)
+            .find(|(_, z)| z.zone_type.is_clickhouse_keeper())
+            .unwrap();
+
+        // Expunge a keeper zone
+        let mut builder = input.into_builder();
+        builder.sleds_mut().get_mut(&sled_id).unwrap().policy =
+            SledPolicy::Expunged;
+        let input = builder.build();
+
+        let blueprint4 = Planner::new_based_on(
+            log.clone(),
+            &blueprint3,
+            &input,
+            "test_blueprint4",
+            &collection,
+        )
+        .expect("created planner")
+        .with_rng_seed((TEST_NAME, "bp4"))
+        .plan()
+        .expect("plan");
+
+        // The planner should expunge a zone based on the sled being expunged. Since this
+        // is a clickhouse keeper zone, the clickhouse keeper configuration should change
+        // to reflect this.
```

Comment:
Just for my own understanding. Let's say we want to move a keeper node (from a cluster or 3) from one zone to another newly created zone. In this scenario we would:

1. Change the XML configuration files to a 2 keeper node cluster and deploy it .
2. Change the XML configuration files to a 3 keeper node cluster (including the host address of the new zone and a new keeper ID) and deploy it.

This means we would not:

1. Change the XML configuration files to a 3 keeper node cluster by removing the old keeper node's host address and ID and adding the new keeper node's host address and ID, and deploying that

Is this accurate?

Reply:
Correct. However, I'm still not sure if we'll end up using file configuration or the `reconfig` command for the keepers. I need to do some more testing once I get to the execution phase.

Reply:
Sadly, it appears that our current version of the keeper cli does not have the `reconfig` command :(

```
$ clickhouse keeper-client -h localhost -p 20001 --q help
cd [path] -- Change the working path (default `.`)
create <path> <value> [mode] -- Creates new node with the set value
delete_stale_backups -- Deletes ClickHouse nodes used for backups that are now inactive
find_big_family [path] [n] -- Returns the top n nodes with the biggest family in the subtree (default path = `.` and n = 10)
find_super_nodes <threshold> [path] -- Finds nodes with number of children larger than some threshold for the given path (default `.`)
flwc <command> -- Executes four-letter-word command
get <path> -- Returns the node's value
get_stat [path] -- Returns the node's stat (default `.`)
help -- Prints this message
ls [path] -- Lists the nodes for the given path (default: cwd)
rm <path> -- Remove the node
rmr <path> -- Recursively deletes path. Confirmation required
set <path> <value> [version] -- Updates the node's value. Only update if version matches (default: -1)
touch <path> -- Creates new node with an empty string as value. Doesn't throw an exception if the node already exists

$ clickhouse keeper-client -h localhost -p 20001 --q reconfig
Syntax error: failed at position 1 ('reconfig'):

reconfig

Expected one of: Keeper client query, cd, create, delete_stale_backups, find_big_family, find_super_nodes, flwc, get, get_stat, help, ls, rm, rmr, set, touch, lgif, csnp, dump, wchp, rqld, wchc, isro, crst, dirs, cons, srst, envi, conf, stat, ruok, srvr, wchs, mntr
```

We may have to use the configuration files for now until we're able to get onto a newer version

Reply:
Good to know. Thanks for pointing this out @karencfv.